src/Microsoft.OData.Core/ODataServiceCollectionExtensions.cs

```diff
@@ -55,15 +55,15 @@ public static IServiceCollection AddDefaultODataServices(this IServiceCollection
             // Finally, register configurable settings.
             var readerSettings = new ODataMessageReaderSettings(odataVersion);
             configureReaderAction?.Invoke(readerSettings);
-            services.AddScoped(sp => readerSettings);
+            services.AddScoped(sp => readerSettings.Clone());
 
             var writerSettings = new ODataMessageWriterSettings(odataVersion);
             configureWriterAction?.Invoke(writerSettings);
-            services.AddScoped(sp => writerSettings);
+            services.AddScoped(sp => writerSettings.Clone());

```

Comment:
When an `ODataMessageWriterSettings` is passed as argument to an `ODataMessageWriter`, the scoped `ODataMessageWriterSettings` service is being updated to that value?

Reply:
That's a very question, and after checking the code, apparently we do :(. The settings from the `ODataMessageWriterSettings` are copied to the scoped `ODataMessageWriterSettings` service:

see:
- https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/ODataMessageWriter.cs#L110
- https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/ODataMessageWriterSettings.cs#L485

which calls:

```c#
internal static ODataMessageWriterSettings CreateWriterSettings(
    IServiceProvider container,
    ODataMessageWriterSettings other)
{
    ODataMessageWriterSettings writerSettings;
    if (container == null)
    {
        writerSettings = new ODataMessageWriterSettings();
    }
    else
    {
       // we get the scoped settings
        writerSettings = container.GetRequiredService<ODataMessageWriterSettings>();
    }

    if (other != null)
    {
        // then overwrite the scoped service based on the settings passed to the writer
        writerSettings.CopyFrom(other);
    }

    return writerSettings;
}
```

Reply:
That explains the observed bug and confirms the hypothesis: the same settings instance is indeed re-used and updated concurrently.
Your change will fix it.